test/Maps/MapWriter.test.cpp

```diff
@@ -0,0 +1,47 @@
+#include "Maps/MapWriter.h"
+#include "Streams/MemoryWriter.h"
+#include "XFile.h"
+#include <gtest/gtest.h>
+#include <vector>
+#include <stdexcept>
+
+TEST(MapWriter, EmptyMapData)
+{
+	MapData mapData;
+	
+	// Write to File
+	const std::string testFilename("maps/data/test.map");
+	EXPECT_THROW(MapWriter::Write(testFilename, mapData), std::runtime_error);
+	XFile::DeletePath(testFilename);
+
+	// Write to Memory
+	std::vector<char> buffer;
+	buffer.resize(1000);
+	Stream::MemoryWriter memoryWriter(buffer.data(), buffer.size() * sizeof(decltype(buffer)::value_type));
+	EXPECT_THROW(MapWriter::Write(memoryWriter, mapData), std::runtime_error);
+}
+
+TEST(MapWriter, BlankFilename)
+{
+	MapData mapData;
+
+	EXPECT_THROW(MapWriter::Write("", mapData), std::runtime_error);
+}
+
+TEST(MapWriter, VersionTag) {
+	const std::string testFilename("maps/data/test.map");
+	MapData mapData;
+	const int minVersionTag = 0x1010;
```

Comment:
This should probably become a public member of the `MapData` class. We should also perhaps add a note that the game itself checks for this minimum version, and will fail to load a game otherwise.

Reply:
I see some oddities with making this constant a static member of the `MapData`. The constant is used in `MapHeader`, which is in a separate header file included by `MapData`. There would be a circular dependency. Putting the constant in `MapHeader` would avoid the circular dependency, but doesn't seem to make as much sense.

Perhaps `MapHeader` isn't really an independent struct. Perhaps it should be declared within the `MapData` struct. The dependencies would make more sense that way. It would allow the constant to be declared in a place accessible to both the parent struct and the sub struct.

Reply:
Additionally, we would need to *define* the constant in a .cpp file. We can *declare* and set an initial value in a header file. That would be sufficient for some uses. But if we ever take the address of the constant, it would need to be defined in a .cpp file or we'd get errors.

Declaration (in .h):
```cpp
static const int32_t MinimumSupportedVersion = 0x1010;
```

Definition (in .cpp):
```cpp
const int32_t MapData::MinimumSupportedVersion;
```

When writing the map data, it may be useful to take the address of this constant to pass to a Stream::Write method. Hence I'm inclined to put the definition there.

----

The placement of the constant definition makes me think the reading and writing code is not very independent of the host struct. Perhaps it should not be separated into independent files. Indeed, the structure of these files is noticeably different from other classes and struct in the project.

Reply:
I think we should keep the MapHeader struct. However, I think it should be used as a local variable during saving and loading, rather than being a member variable of MapData. Some fields are worth saving as members, some aren't.